### PR TITLE
Format <foo + 'bar> nicely

### PR DESCRIPTION
**Pull request by nrc**

### CODE REVIEW

File: src/types.rs [hunk 1]
```diff
@@ -124,19 +123,19 @@ impl<'a> SegmentParam<'a> {
...
-            }
-            SegmentParam::Type(ref ty) => {
-                write!(f, "{}", pprust::ty_to_string(ty))
-            }
-            SegmentParam::Binding(ref binding) => {
-                write!(f, "{} = {}", binding.ident, pprust::ty_to_string(&*binding.ty))
-            }
-        }
+impl<'a> Rewrite for SegmentParam<'a> {
+    fn rewrite(&self, context: &RewriteContext, width: usize, offset: usize) -> Option<String> {
```
Comment (marcusklaas) [inline src/types.rs]: This doesn't uphold the variant that the single-line result of a rewrite does not exceed `width` characters, which could lead to Bad Stuff..

File: src/types.rs [hunk 2]
```diff
@@ -346,6 +348,13 @@ impl Rewrite for ast::TyParamBound {
     }
 }
 
+impl Rewrite for ast::TyParamBounds {
+    fn rewrite(&self, context: &RewriteContext, width: usize, offset: usize) -> Option<String> {
```
Comment (marcusklaas) [inline src/types.rs]: Ditto.

File: src/types.rs [hunk 3]
```diff
@@ -389,3 +398,19 @@ impl Rewrite for ast::PolyTraitRef {
         }
     }
 }
+
+impl Rewrite for ast::Ty {
```
Comment (marcusklaas) [inline src/types.rs]: Idem.

**Comment by marcusklaas** [comment]
There are still some invocation of `ty_to_string` in `items.rs`, should we replace those as well?

**Comment by marcusklaas** [comment]
These are good changes, but it'd be nice to uphold the "rewrite invariant". We can just add some FIXMEs and do this later, too.

**Comment by nrc** [comment]
Yeah, it's a shame not to maintain the rewrite invariants, but I don't think I can do that without completely implementing formatting types, and I don't really want to do that - there are other higher priority things to look at.

I'll add some fixmes and replace the ty_to_strings in items.rs

**Comment by nrc** [comment]
Actually some of the occurrences of ty_to_string I've left since it wasn't easy to get a context

**Comment by marcusklaas** [comment]
Can't we make the string, compare its length to width and return `None` if it's greater?

-----Original Message-----

**Comment by nrc** [comment]
We could, and we should some day, but I worry that the callers won't do the work to reformat, and even calculating the available width etc. for those calls is hard work.

**Pull request merged**